diff --git a/ggml/src/ggml-cpu/llamafile/sgemm.cpp b/ggml/src/ggml-cpu/llamafile/sgemm.cpp
index 5efaaa5b2..2b285514a 100644
--- a/ggml/src/ggml-cpu/llamafile/sgemm.cpp
+++ b/ggml/src/ggml-cpu/llamafile/sgemm.cpp
@@ -1542,11 +1542,26 @@ class tinyBLAS_Q0_AVX {
                 __m256i avec1 = load(A + lda * (ii + 1) + l);
                 __m256i avec2 = load(A + lda * (ii + 2) + l);
                 __m256i avec3 = load(A + lda * (ii + 3) + l);
+                __m256 dvec2[RN];
+                if constexpr (RN == 2 && std::is_same_v<TA, block_q8_0> && std::is_same_v<TB, block_q8_0>) {
+                    const uint32_t b_delta = B[ldb * (jj + 0) + l].d |
+                                             ((uint32_t) B[ldb * (jj + 1) + l].d << 16);
+                    const __m128 db = _mm_cvtph_ps(_mm_cvtsi32_si128(b_delta));
+                    dvec2[0] = _mm256_castps128_ps256(_mm_mul_ps(da, _mm_shuffle_ps(db, db, 0)));
+                    dvec2[0] = _mm256_permute2f128_ps(dvec2[0], dvec2[0], 0);
+                    dvec2[1] = _mm256_castps128_ps256(_mm_mul_ps(da, _mm_shuffle_ps(db, db, 85)));
+                    dvec2[1] = _mm256_permute2f128_ps(dvec2[1], dvec2[1], 0);
+                }
                 for (int64_t j = 0; j < RN; ++j) {
-                        __m128 db = _mm_set1_ps(unhalf(B[ldb * (jj + j) + l].d));
                         // Computation of product of delta values for four blocks and replicate it across 256 bit lane
-                        __m256 dvec =  _mm256_castps128_ps256(_mm_mul_ps(da, db));
-                        dvec = _mm256_permute2f128_ps(dvec ,dvec, 0);
+                        __m256 dvec;
+                        if constexpr (RN == 2 && std::is_same_v<TA, block_q8_0> && std::is_same_v<TB, block_q8_0>) {
+                            dvec = dvec2[j];
+                        } else {
+                            __m128 db = _mm_set1_ps(unhalf(B[ldb * (jj + j) + l].d));
+                            dvec = _mm256_castps128_ps256(_mm_mul_ps(da, db));
+                            dvec = _mm256_permute2f128_ps(dvec, dvec, 0);
+                        }
                         // Computation of dot product and multiplication with appropriate delta value products
                         Cv[j][0] = madd(_mm256_shuffle_ps(dvec, dvec, 0),
                                     updot(_mm256_sign_epi8(avec0, avec0),
